hypothesis-python/src/hypothesis/searchstrategy/attrs.py

```diff
@@ -0,0 +1,120 @@
+# coding=utf-8
+#
+# This file is part of Hypothesis, which may be found at
+# https://github.com/HypothesisWorks/hypothesis-python
+#
+# Most of this work is copyright (C) 2013-2018 David R. MacIver
+# (david@drmaciver.com), but it contains contributions by others. See
+# CONTRIBUTING.rst for a full list of people who may hold copyright, and
+# consult the git log if you need to determine who owns an individual
+# contribution.
+#
+# This Source Code Form is subject to the terms of the Mozilla Public License,
+# v. 2.0. If a copy of the MPL was not distributed with this file, You can
+# obtain one at http://mozilla.org/MPL/2.0/.
+#
+# END HEADER
+
+from __future__ import division, print_function, absolute_import
+
+from functools import reduce
+from itertools import chain
+from collections import defaultdict
+
+import attr
+
+import hypothesis.strategies as st
+from hypothesis.errors import ResolutionFailed
+from hypothesis.internal.compat import get_type_hints
+from hypothesis.utils.conventions import infer
+
+
+def from_attrs(target, args, kwargs, to_infer):
+    fields = attr.fields(target)
+    kwargs = {k: v for k, v in kwargs.items() if v is not infer}
+    for name in to_infer:
+        kwargs[name] = from_attrs_attribute(getattr(fields, name))
+    # TODO: add a layer here that retries drawing if validation fails, for
+    # improved composition.  Precedent: timezones for datetimes().
+    return st.tuples(st.tuples(*args), st.fixed_dictionaries(kwargs)).map(
+        lambda value: target(*value[0], **value[1])
+    )
+
+
+def from_attrs_attribute(attrib):
+    """Infer a strategy from an attr.Attribute object."""
```

Comment:
I think I need some more high level commenting about what's going here, separate from the code. I'm not having a very awake day, but I'm really struggling to follow the logic here.